[AArch64][Codegen]Transform saturating smull to sqdmulh

nasherm opened the pull request:
This patch adds a pattern for recognizing saturating vector
smull. Prior to this patch these were performed using a
combination of smull+smull2+uzp+smin  like the following

```
        smull2  v5.2d, v1.4s, v2.4s
        smull   v1.2d, v1.2s, v2.2s
        uzp2    v1.4s, v1.4s, v5.4s
        smin    v1.4s, v1.4s, v0.4s
```

which now optimizes to

```
    sqdmulh v0.4s, v1.4s, v0.4s
```

This only operates on vectors containing int32 and int16 types

nasherm committed:
[AArch64][Codegen]Transform saturating smull to sqdmulh

This patch adds a pattern for recognizing saturating vector
smull. Prior to this patch these were performed using a
combination of smull+smull2+uzp+smin  like the following

```
        smull2  v5.2d, v1.4s, v2.4s
        smull   v1.2d, v1.2s, v2.2s
        uzp2    v1.4s, v1.4s, v5.4s
        smin    v1.4s, v1.4s, v0.4s
        add     v1.4s, v1.4s, v1.4s
```

which now optimizes to

```
    sqdmulh v0.4s, v1.4s, v0.4s
    sshr    v0.4s, v0.4s, #1
    add v0.4s, v0.4s, v0.4s
```

This only operates on vectors containing Q31 data types.

Change-Id: Ib7d4d5284d1bd3fdd0907365f9e2f37f4da14671

NickGuy-Arm commented on llvm/lib/Target/AArch64/AArch64InstrInfo.td:
```diff
@@ -9349,6 +9349,13 @@ def : Pat<(v4i32 (mulhs V128:$Rn, V128:$Rm)),
+// Saturating vector mulitplications on signed integers
+// follow a smull + smull2 + uzip + smin pattern. It would 
+// be more efficient to make use of sqdmulh instructions
+def : Pat<(v4i32 (smin (mulhs V128:$Rn, V128:$Rm), 
+                        (v4i32 (AArch64mvni_shift (i32 192), (i32 24))))),
+           (USHRv4i32_shift (SQDMULHv4i32 V128:$Rn, V128:$Rm), (i32 1))>;
```
Could we support more than only `v4i32`, and maybe scalable vectors?

nasherm replied:
I'm not 100% sure. I'd have to have a think. I know that SQDMULH only supports v8i16 and v4i32. SVE I'm unsure of.

NickGuy-Arm reviewed (commented):
Could we get a test precommit to see the changes in action?

nasherm commented on llvm/lib/Target/AArch64/AArch64InstrInfo.td:
```diff
@@ -9349,6 +9349,13 @@ def : Pat<(v4i32 (mulhs V128:$Rn, V128:$Rm)),
+// be more efficient to make use of sqdmulh instructions
+def : Pat<(v4i32 (smin (mulhs V128:$Rn, V128:$Rm), 
+                        (v4i32 (AArch64mvni_shift (i32 192), (i32 24))))),
+           (USHRv4i32_shift (SQDMULHv4i32 V128:$Rn, V128:$Rm), (i32 1))>;
```
I've updated my patch to make use of AArch64ISelLowering to make it more generic

nasherm commented:
Ping

I have made changes to AArch64ISelLowering to enable this transform for `v4i32` types if someone could take a look? I believe this can be extended to `v8i16` types but would require a bit more work as clang generates a different IR pattern when the type is `v8i16`. 

As for SVE I haven't had bandwidth to look at enabling this so I don't currently know if this would be possible/beneficial for an SVE equivalent pattern

davemgreen commented:
Hi - sorry for the delay, I was trying to re-remember how this instruction worked. This has three functions: https://godbolt.org/z/cxb7osTP1, the first of which I think is the most basic for of sqdmulh (notice the >2x bitwidth extend to allow the mul and the x2 to not wrap, and the min+max to saturate). That is equivalent to the @updated which is what llvm will optimize it to (the mul x2 is folded into the shift, and the only value that can actually saturate is -0x8000*-0x8000). It is equivalent the the third I believe because we only need 2x the bitwidth in this form.

That feels like the most basic form of sqdmulh. Any reason not to add that one first? I didn't look into this pattern a huge amount, but do you know if the bottom or top bits require the shifts? Or both?

nasherm committed:
Responding to review comments

Based on the most recent PR comments I've
- refactored the change to work on a reduced pattern
  which is truer to the actual SQDMULH instruction
- written pattern matches for q31, q15 and int32, int16
  data types
- rewritten and extended the tests

Change-Id: I18c05e56b3979b8dd757d533e44a65496434937b

nasherm committed:
Arithmetic error for Q types

Spotted and fixed an artihmetic error when working
with Q types

Change-Id: I80f8e04bca08d3e6bc2740201bdd4978446a397f

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -20717,6 +20717,83 @@ static SDValue performBuildVectorCombine(SDNode *N,
+  SDValue SExt0 = Mul.getOperand(0);
+  SDValue SExt1 = Mul.getOperand(1);
+
+  if (SExt0.getOpcode() != ISD::SIGN_EXTEND ||
+      SExt1.getOpcode() != ISD::SIGN_EXTEND)
```
Check for things like the types of the SExt's match, and that they are the legal types you expect them to be.

It can also be good to test with larger types like <8 x i32>, to see if they match correctly after type legalization. They are often more/different nodes but the important part is that they do not fail to select because we produced illegal nodes.

nasherm committed:
Responding to review comments

- support for v2i32 and v4i16 patterns
- extra type checking on sext
- matching on smin over sext
- cleaning trailing lines

Change-Id: I9f61b8d77a61f3d44ad5073b41555c9ad5653e1a

nasherm commented on llvm/lib/Target/AArch64/AArch64InstrInfo.td:
```diff
@@ -9365,6 +9368,13 @@ def : Pat<(v4i32 (mulhu V128:$Rn, V128:$Rm)),
            (UMULLv4i32_v2i64 V128:$Rn, V128:$Rm))>;
 
+
+def : Pat<(v8i16 (AArch64sqdmulh (v8i16 V128:$Rn), (v8i16 V128:$Rm))),
+          (SQDMULHv8i16 V128:$Rn, V128:$Rm)>;
```
Done

nasherm commented on llvm/test/CodeGen/AArch64/saturating-vec-smull.ll:
```diff
@@ -0,0 +1,58 @@
+define <8 x i16> @saturating_q15(<8 x i16> %a, <8 x i16> %b) {
+; CHECK-LABEL: saturating_q15:
+; CHECK:       // %bb.0:
+; CHECK-NEXT:    sqdmulh v0.8h, v1.8h, v0.8h
```
I think I'd have to add later patches to support Q types now that I think about it. So I've opted for this patch to just focus on supporting int32 and int16

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -20717,6 +20717,83 @@ static SDValue performBuildVectorCombine(SDNode *N,
   return SDValue();
 }
 
+// A special combine for the vqdmulh family of instructions.
```
vqdmulh->sqdmulh

nasherm replied:
Done

davemgreen commented on llvm/test/CodeGen/AArch64/saturating-vec-smull.ll:
```diff
@@ -0,0 +1,81 @@
+; CHECK-NEXT:    sqdmulh v4.2s, v5.2s, v4.2s
+; CHECK-NEXT:    sqdmulh v3.2s, v7.2s, v6.2s
+; CHECK-NEXT:    mov v1.d[1], v4.d[0]
+; CHECK-NEXT:    mov v0.d[1], v3.d[0]
```
This looks OK. If you want to make this better, it might be possible to add a concat(sqdmulh, sqdmulh) -> sqdmulh combine. Look for "Optimise concat_vectors of two identical binops with a 128-bit destination"

nasherm committed:
Responding to review comments

- minor cleanup
- allow optimizing concat_vectors(sqdmulh,sqdmulh) -> sqdmulh
- testing EVTs better

Change-Id: I0404fb9900896050baac372b7f7ce3a5b03517b9

nasherm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -20717,6 +20717,83 @@ static SDValue performBuildVectorCombine(SDNode *N,
+  SDLoc DL(SMin);
+  EVT VecVT = N->getValueType(0);
+  SDValue SQDMULH = DAG.getNode(AArch64ISD::SQDMULH, DL, VecVT, V0, V1);
+  return DAG.getNode(ISD::SIGN_EXTEND, DL, N->getValueType(0), SQDMULH);
```
Done

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -20918,6 +20928,79 @@ static SDValue performBuildVectorCombine(SDNode *N,
+// SATURATING_VAL ) can be reduced to sext(sqdmulh(...))
+static SDValue trySQDMULHCombine(SDNode *N, SelectionDAG &DAG) {
+
+  if (N->getOpcode() != ISD::TRUNCATE)
```
Can it start for a smin.

davemgreen replied:
Yep start from the SMIN, in something like performSMINCombine. (It doesn't exist yet, and performDAGCombine  can call trySQDMULHCombine directly for SMIN).

The idea is not to match `trunc(smin(asr(mul(..))))` and generate `sqdmulh(..)`, but instead to match `smin(asr(mul(..))))` and generate `sext(sqdmulh(..))`, which should always be equivalent but more generic. We might temporarily have `trunc(sext(sqdmulh(..)))`, but the trunc+sext will optimize away.

davemgreen commented on llvm/test/CodeGen/AArch64/saturating-vec-smull.ll:
```diff
@@ -0,0 +1,73 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=aarch64-none-elf < %s | FileCheck %s
+
+define <4 x i16> @saturating_4xi16(<4 x i16> %a, <4 x i16> %b) {
```
Can you add a test case for this, in case it goes wrong.
```
define <2 x i16> @saturating_2xi16(<2 x i16> %a, <2 x i16> %b) {
  %as = sext <2 x i16> %a to <2 x i32>
  %bs = sext <2 x i16> %b to <2 x i32>
  %m = mul <2 x i32> %bs, %as
  %sh = ashr <2 x i32> %m, splat (i32 15)
  %ma = tail call <2 x i32> @llvm.smin.v4i32(<2 x i32> %sh, <2 x i32> splat (i32 32767))
  %t = trunc <2 x i32> %ma to <2 x i16>
  ret <2 x i16> %t
}
```

Also I would add some test cases that were not expected to transform, like shifiting by 14 or having the wrong saturate constant.

davemgreen commented on llvm/test/CodeGen/AArch64/saturating-vec-smull.ll:
```diff
@@ -0,0 +1,73 @@
+  ret <8 x i16> %t
+}
+
+define <2 x i32> @saturating_2xi32(<2 x i32> %a, <2 x i32> %b) {
```
Also can you add a test like this, to capture starting from the smin:
```
define <2 x i64> @saturating_2xi32(<2 x i32> %a, <2 x i32> %b) {
  %as = sext <2 x i32> %a to <2 x i64>
  %bs = sext <2 x i32> %b to <2 x i64>
  %m = mul <2 x i64> %bs, %as
  %sh = ashr <2 x i64> %m, splat (i64 31)
  %ma = tail call <2 x i64> @llvm.smin.v8i64(<2 x i64> %sh, <2 x i64> splat (i64 2147483647))
  ret <2 x i64> %ma
}
```
It should just be `sshll(sqdmulh(v0, v1))`

nasherm replied:
Done

nasherm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -20918,6 +20928,79 @@ static SDValue performBuildVectorCombine(SDNode *N,
+  }
+
+  SDValue Sra = SMin.getOperand(0);
+  if (Sra.getOpcode() != ISD::SRA)
```
Done

nasherm committed:
Responding to review comments

- making sure transform only operates on smin nodes
- adding extra tests dealing with interesting edge cases

Change-Id: Ia1114ec9b93c4de3552b867e0d745beccdae69f1

nasherm commented:
There's a rebase error in my most recent patch that I need to fix

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -20918,6 +20929,90 @@ static SDValue performBuildVectorCombine(SDNode *N,
+  SDValue SExt0 = Mul.getOperand(0);
+  SDValue SExt1 = Mul.getOperand(1);
+
+  EVT SExt0Type = SExt0.getOperand(0).getValueType();
+  EVT SExt1Type = SExt1.getOperand(0).getValueType();
```
It is probably best to check the opcode type before calling getOperand(0) on them. It could be something like POISON.

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -20918,6 +20929,90 @@ static SDValue performBuildVectorCombine(SDNode *N,
+
+  // Source vectors with width < 64 are illegal and will need to be extended
+  unsigned SourceVectorWidth = SExt0Type.getFixedSizeInBits();
+  SDValue V0 = (SourceVectorWidth < 64) ? SExt0 : SExt0.getOperand(0);
+  SDValue V1 = (SourceVectorWidth < 64) ? SExt1 : SExt1.getOperand(0);
```
I'm not sure why this is valid to use SExt0 directly? Could it check types directly and produce `sqdmul(sext(SExt0.getOperand(0)), sext(SExt0.getOperand(0)))` if needed? Where the sext is always to a legal type. What happens at the moment with a v2i48?

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -20918,6 +20929,90 @@ static SDValue performBuildVectorCombine(SDNode *N,
+  SDValue SQDMULH =
+      DAG.getNode(AArch64ISD::SQDMULH, DL, V0.getValueType(), V0, V1);
+  EVT DestVT = N->getValueType(0);
+  if (DestVT.getScalarSizeInBits() > SExt0Type.getScalarSizeInBits())
+    return DAG.getNode(ISD::SIGN_EXTEND, DL, DestVT, SQDMULH);
```
I think it should always return `sext(SQDMULH)` or else the transform wasn't valid (there were not enough bits to prevent wrapping).

nasherm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -20918,6 +20929,90 @@ static SDValue performBuildVectorCombine(SDNode *N,
+
+  EVT VT = N->getValueType(0);
+
+  if (!VT.isVector() || VT.getScalarSizeInBits() > 64)
```
Done

nasherm committed:
Response to review comments addressing:

- check for scalar type
- check for sign extends
- legalise vector inputs to sqdmulh
- always return sext(sqdmulh)

Change-Id: Ic58b7f267e94bc2592942fc29b829ffb6221770f

nasherm committed:
Clang format

Change-Id: Ifeba39acab171c75df496f89688fda701bd7dd85

davemgreen reviewed (commented):
Can you add this test case and fix what is wrong with it? Sometimes it is best to just check for all the types that are valid.
```
define <6 x i16> @saturating_6xi16(<6 x i16> %a, <6 x i16> %b) {
  %as = sext <6 x i16> %a to <6 x i32>
  %bs = sext <6 x i16> %b to <6 x i32>
  %m = mul <6 x i32> %bs, %as
  %sh = ashr <6 x i32> %m, splat (i32 15)
  %ma = tail call <6 x i32> @llvm.smin.v6i32(<6 x i32> %sh, <6 x i32> splat (i32 32767))
  %t = trunc <6 x i32> %ma to <6 x i16>
  ret <6 x i16> %t
}
```

nasherm committed:
Review comments

- added testcase for v6i16  and fixed issues
- added testcases for v1i16  and fixed issues

Change-Id: I4694c48ff9f12ee6048efd2394d5b710df7ebbea

davemgreen reviewed (approved):
Thanks. LGTM
